core/services/relay/evm/event_binding.go

```diff
@@ -0,0 +1,270 @@
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+)
+
+// Max four topics on EVM, the first topic is always the event signature, so 3 topics left for fields
+const maxTopicFields = 3
+
+type eventBinding struct {
+	address        common.Address
+	contractName   string
+	eventName      string
+	lp             logpoller.LogPoller
+	hash           common.Hash
+	codec          commontypes.RemoteCodec
+	pending        bool
+	bound          bool
+	registerCalled bool
+	lock           sync.Mutex
+	inputInfo      *codecEntry
+	inputModifier  codec.Modifier
+	topicInfo      *codecEntry
+	// used to allow Register and Unregister to be unique in case two bindings have the same event.
+	// otherwise, if one unregisters, it'll unregister both with the LogPoller.
+	id string
+}
+
+var _ readBinding = &eventBinding{}
+
+func (e *eventBinding) SetCodec(codec commontypes.RemoteCodec) {
+	e.codec = codec
+}
+
+func (e *eventBinding) Register() error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	e.registerCalled = true
+	if !e.bound || e.lp.HasFilter(e.id) {
+		return nil
+	}
+
+	if err := e.lp.RegisterFilter(logpoller.Filter{
+		Name:      e.id,
+		EventSigs: evmtypes.HashArray{e.hash},
+		Addresses: evmtypes.AddressArray{e.address},
+	}); err != nil {
+		return fmt.Errorf("%w: %w", commontypes.ErrInternal, err)
+	}
+	return nil
+}
+
+func (e *eventBinding) Unregister() error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	if !e.lp.HasFilter(e.id) {
+		return nil
+	}
+
+	if err := e.lp.UnregisterFilter(e.id); err != nil {
+		return fmt.Errorf("%w: %w", commontypes.ErrInternal, err)
+	}
+	return nil
+}
+
+func (e *eventBinding) GetLatestValue(ctx context.Context, params, into any) error {
+	if !e.bound {
+		return fmt.Errorf("%w: event not bound", commontypes.ErrInvalidType)
+	}
+
+	confs := logpoller.Finalized
+	if e.pending {
+		confs = logpoller.Unconfirmed
+	}
+
+	if len(e.inputInfo.Args) == 0 {
+		return e.getLatestValueWithoutFilters(ctx, confs, into)
+	}
+
+	return e.getLatestValueWithFilters(ctx, confs, params, into)
+}
+
+func (e *eventBinding) Bind(binding commontypes.BoundContract) error {
+	if err := e.Unregister(); err != nil {
+		return err
+	}
+
+	e.address = common.HexToAddress(binding.Address)
+	e.pending = binding.Pending
+	e.bound = true
+
+	if e.registerCalled {
+		return e.Register()
+	}
+	return nil
+}
+
+func (e *eventBinding) getLatestValueWithoutFilters(ctx context.Context, confs logpoller.Confirmations, into any) error {
+	log, err := e.lp.LatestLogByEventSigWithConfs(e.hash, e.address, confs)
+	if err = wrapInternalErr(err); err != nil {
+		return err
+	}
+
+	return e.decodeLog(ctx, log, into)
+}
+
+func (e *eventBinding) getLatestValueWithFilters(
+	ctx context.Context, confs logpoller.Confirmations, params, into any) error {
+	offChain, err := e.convertToOffChainType(params)
```

Comment:
can you remind me how this works? For eg. Product X wants to `getLatestValue`  of some event by an indexed topic, lets say address(probably too simple of an example). What is the corresponding OffChainType to the type that the product passed

Reply:
Probably something like

type Args {
 Address []byte
}

the `checkedType` and `nativeType` would be `common.Address` and would assure that it's the right size.  `decodeAccountHook` allows the passed type to also be string like "0x1234" for hard-coding in config, since those aren't retrieved from somewhere and are human written.